File: 3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring.py
class lengthOfLongestSubstring:
    def solution(self, s: str) -> int:
        if len(s) == 0:
            return 0
        left = 0
        curCnt = 0
        maxCnt = 0
        charSet = set()
        for i in range(len(s)):
            curCnt = curCnt + 1
            while s[i] in charSet:
                charSet.remove(s[left])
                left = left + 1
                curCnt = curCnt - 1
            maxCnt = max(maxCnt, curCnt)
            charSet.add(s[i])
        return maxCnt

File: 3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring.java
class lengthOfLongestSubstring {
    public int solution(String s) {
        if (s.length() == 0) {
            return 0;
        }
        HashMap<Character, Integer> charMap = new HashMap<Character, Integer>();
        int left = 0, maxCnt = 0;
        for (int i = 0; i < s.length(); i ++) {
            if (charMap.containsKey(s.charAt(i))) {
                left = Math.max(left, charMap.get(s.charAt(i)) + 1);
            }
            charMap.put(s.charAt(i), i);
            maxCnt = Math.max(maxCnt, i - left + 1);
        }
        return maxCnt;
    }
}